### PR TITLE
vNEXT: SDK UX BoundingBoxRig & TwoHandManipulation

### PR DESCRIPTION
**Pull request by johnppella**
## Overview
Ported [BoundingBoxRig ](https://github.com/Microsoft/MixedRealityToolkit-Unity/blob/master/Assets/HoloToolkit-Examples/UX/Readme/README_BoundingBoxGizmoExample.md)and [TwoHandManipulation ](https://github.com/Microsoft/MixedRealityToolkit-Unity/blob/master/Assets/HoloToolkit/Input/Scripts/Utilities/Interactions/TwoHandManipulatable.cs)from HoloToolkit.
no major logic changes - mainly namespace changes.

New location of the ported sample scenes:
MixedRealityToolkit-**Examples**/Demos/UX/BoundingBox/Scenes/BoundingBoxExample.unity
MixedRealityToolkit-**Examples**/Demos/Input/TwoHandManipulation/Scenes/TwoHandManipulationExample.unity

### InputSystem
Assets/MixedRealityToolkit/InputSystem/Utilities/AffineConstraintEnums.cs

### SDK/Boundnig Box 
**Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/**
BoundingBox.cs
BoundingBoxGizmo.cs
BoundingBoxGizmoHandle.cs
BoundingBoxGizmoHandleAxisToAffectEnum.cs
BoundingBoxGizmoHandleHandMotionTypeEnum.cs
BoundingBoxGizmoHandleRotationTypeEnum.cs
BoundingBoxGizmoHandleTransformTypeEnum.cs
BoundingBoxes/BoundingBoxGizmoShell.cs
BoundingBoxHelper.cs
BoundingBoxRig.cs

**Assets/MixedRealityToolkit-SDK/Features/UX/Materials/**
BoundingBoxHandle.mat
BoundingBoxHandleGrabbed.mat
BoundingBoxHandleMaterial.mat
BoundingBoxLines.mat

### SDK/TwoHandManipulation
**Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/TwoHandManipulation/**
ManipulationMode.cs
TwoHandManipulatable.cs
TwoHandMoveLogic.cs
TwoHandRotateLogic.cs
TwoHandScaleLogic.cs

### Examples/Demos
Demos/Input/TwoHandManipulation/Scenes/TwoHandManipulationExample.unity
Demos/UX/BoundingBox/Scenes/BoundingBoxExample.unity
Demos/UX/Materials/Coffee.mat
Demos/UX/Materials/CoffeeCup.mat
Demos/UX/Models/Buttons.FBX
Demos/UX/Models/CoffeeCup.fbx
Demos/UX/Models/Model_Bucky.fbx
Demos/UX/Models/Model_Icosa.fbx
Demos/UX/Models/Model_Octa.fbx
Demos/UX/Models/Model_Platonic.fbx
Demos/UX/Models/Model_PushButton.fbx
Demos/UX/Models/MouseandCheese.fbx
Demos/UX/Models/TheModule.fbx
Demos/UX/Models/balloon.fbx
Demos/UX/Textures/BoxGrid.png
Demos/UX/Textures/CoffeeFoam.png

### CODE REVIEW

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBox.cs [hunk 1]
```diff
@@ -0,0 +1,640 @@
...
+    /// </summary>
+    public class BoundingBox : MonoBehaviour
+    {
+        public enum BoundsCalculationMethodEnum
+        {
+            MeshFilterBounds,   // Better for flattened objects - this mode also treats RectTransforms as quad meshes
+            RendererBounds,     // Better for objects with non-mesh renderers
+            Colliders,          // Better if you want precise control
+            Default,            // Use the default method (RendererBounds)
+        }
```
Comment (radicalad) [inline Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBox.cs]: Pull out your enums into separate files.

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBox.cs [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+                    flattenedAxis = value;
+                    if (OnFlattenedAxisChange != null)
+                    {
+                        OnFlattenedAxisChange();
+                    }
+                }
+            }
+        }
+
+        #region
```
Comment (radicalad) [inline Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBox.cs]: label your regions.

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBox.cs [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+
+            for (int i = 0; i < boundsPoints.Count; ++i)
+            {
+                boundsPoints[i] = target.transform.localToWorldMatrix.MultiplyPoint(boundsPoints[i]);
+            }
+        }
+
+        private static Vector3[] corners = null;
+        private static Vector3[] rectTransformCorners = new Vector3[4];
+        #endregion
```
Comment (radicalad) [inline Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBox.cs]: I used these static methods all the time in HTK. You should put them into a static extensions class. So helpful for creating bounds based on an object (and its children).

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandleAxisToAffectEnum.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+namespace Microsoft.MixedReality.Toolkit.SDK.UX
+{
+    public enum BoundingBoxGizmoHandleAxisToAffect
+    {
+        X,
```
Comment (radicalad) [inline Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandleAxisToAffectEnum.cs]: initialize your enums with = 0.

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.SDK.UX
+{
+    /// <summary>
+    /// Logic for the gizmo handles in Bounding Box
+    /// </summary>
+    public class BoundingBoxGizmoHandle : MonoBehaviour, IMixedRealityInputHandler, IMixedRealitySourceStateHandler
+    {
+        private BoundingBoxRig rig;
```
Comment (radicalad) [inline Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs]: missing `<summary>` tags

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+            else
+            {
+                cachedRenderer.sharedMaterial = Rig.RotateHandleMaterial;
+            }
+
+            InputSystem.PopModalInputHandler();
+            Rig.FocusOnHandle(null);
+        }
+
+        #region Event Handlers
```
Comment (radicalad) [inline Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs]: make sure to add `/// <inheritdoc />` on interface implementations

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandleAxisToAffectEnum.cs (hunk 4 above)
Comment (radicalad) [reply Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandleAxisToAffectEnum.cs]: I'd also stick all your enums in a /BoundingBoxEnums/ folder.

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandleHandMotionTypeEnum.cs [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+namespace Microsoft.MixedReality.Toolkit.SDK.UX
+{
+    public enum BoundingBoxGizmoHandleHandMotionType
+    {
+        handRotatesToRotateObject,
+        handMovesToRotateObject
+    }
+}
```
Comment (radicalad) [inline Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandleHandMotionTypeEnum.cs]: I believe the file name suffix needs to match `type` as well.

**Comment by cre8ivepark** [comment]
@StephenHodgson Added the list of the change files instead of separating PR. I found that we only have one file in Core(InputSystem).

**Comment by StephenHodgson** [comment]
Looks like we have duplicates.

image

**Comment by StephenHodgson** [comment]
File and Type names should be the same. Also, we already have this enum defined.

image

**Comment by StephenHodgson** [comment]
We need to also disable the default material imports on the models in Demos/UX/Models.

**Comment by StephenHodgson** [comment]
I also noticed there was a prefab in the models folder. Let's pull the prefabs out into their own.
image

**Comment by StephenHodgson** [comment]
Here's a PR with a few of the changes I noticed we can fix right away:https://github.com/johnppella/MixedRealityToolkit-Unity/pull/1

**Comment by StephenHodgson** [comment]
Seems like most of the code in `BoundingBoxHelper` can be moved to the extension classes.

**Comment by StephenHodgson** [comment]
Also, did we make sure that the meta files where changed from the HTK to the MRTK?

**Comment by cre8ivepark** [comment]
> File and Type names should be the same. Also, we already have this enum defined.
> 
> image

@StephenHodgson it seems like other existing files are already using AxisConstraint including ControllerPopupWindow, MixedRealityGestureProfileInspector, MixedRealityInputActionsProfileInspector. Do we want to update all of these to use RotationConstraintType?

**Comment by StephenHodgson** [comment]
@cre8ivepark I think it's generic enough. I already made the change in the PR John merged in

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBox.cs (hunk 1 above)
Comment (johnppella) [reply Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBox.cs]: yes

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/FlattenModeEnum.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+namespace Microsoft.MixedReality.Toolkit.SDK.UX
+{
+    /// <summary>
+    /// Flatten behavior when object is surrounded by BoundingBoxRig.
+    /// </summary>
+    public enum FlattenModeEnum
+    {
+        DoNotFlatten,   // Always use XYZ axis
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/FlattenModeEnum.cs]: can we make these summary tags?

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundsCalculationMethodEnum.cs [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+namespace Microsoft.MixedReality.Toolkit.SDK.UX
+{
+    /// <summary>
+    /// which method to use to calculate bounds since not all objects have colliders.
+    /// </summary>
+    public enum BoundsCalculationMethodEnum
+    {
+        MeshFilterBounds,   // Better for flattened objects - this mode also treats RectTransforms as quad meshes
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundsCalculationMethodEnum.cs]: we also need to initialize our enums 
Ex:
`MeshFilterBounds = 0,`

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandleAxisToAffectEnum.cs (hunk 4 above)
Comment (johnppella) [reply Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandleAxisToAffectEnum.cs]: they all seem specialized except AxisToAffect

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandleHandMotionTypeEnum.cs (hunk 7 above)
Comment (johnppella) [reply Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandleHandMotionTypeEnum.cs]: this file has been deleted.

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/FlattenModeEnum.cs (hunk 8 above)
Comment (johnppella) [reply Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/FlattenModeEnum.cs]: yes

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxEnums/BoundingBoxGizmoHandleAxisToAffectEnum.cs [hunk 10]
```diff
@@ -3,9 +3,12 @@
 
 namespace Microsoft.MixedReality.Toolkit.SDK.UX
 {
+    /// <summary>
+    /// Which axis does a BoundingBoxRigGizmoHandle rotate around?
+    /// </summary>
     public enum BoundingBoxGizmoHandleAxisToAffect
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxEnums/BoundingBoxGizmoHandleAxisToAffectEnum.cs]: This seems really generic, can we possibly move this into core?

**Comment by johnppella** [comment]
code changes are checked in.

File: Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxEnums/BoundingBoxGizmoHandleAxisToAffectEnum.cs (hunk 10 above)
Comment (johnppella) [reply Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxEnums/BoundingBoxGizmoHandleAxisToAffectEnum.cs]: sure! where in Core would be the best place?

Comment (johnppella) [reply Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/BoundingBoxes/BoundingBoxEnums/BoundingBoxGizmoHandleAxisToAffectEnum.cs]: I moved this enum file to Core->Definitions->Utilities and renamed CardinalAxisType

**Comment by StephenHodgson** [comment]
Good changes. I like where this is going!

**Comment by SimonDarksideJ** [comment]
Any update on this @johnppella 
Like with the other UX PR's, it would be better to break this up so it's easier to review.

**Comment by johnppella** [comment]
IM going to close this PR since the Boundingbox Rig is being re-written to be MUCH simpler.

**Pull request closed without merging**